### ISSUE TITLE
How to handle reporting of progress of asynchronous tasks over ReST?

### ISSUE THREAD

**Issue by jmkhael** [zalando/restful-api-guidelines#559]
First thank you for these very nice and reusable guidelines and tooling (zally).

In the context of tracking progress of long running jobs in asynchronous way, we inspired from 
[restalk operation polling](http://restalk-patterns.org/long-running-operation-polling.html)

This suggests that the GET on an not ready resource (still processing) would return a 200 when not ready and will return a 303 once the resource is ready.

However, the [redirection codes guidelines](https://opensource.zalando.com/restful-api-guidelines/#redirection-codes) do not document(forbids?) that the `GET` do not return a 303.

I was curious how do you handle reporting the progress of an async tasks over ReST?

**Comment by tfrauenstein** [comment]
Thx for reaching out. We currently do not (yet) have explicit guidelines for asynchronous job execution. With respect to the link you provided, IMO there is a difference between the job and the job result. I would expect that the job resource has state e.g. the processing status (and processing progress etc.) and that this it is returned to client as payload, instead of redirecting it to a new job result / output resource once the job is finished.

**Comment by whiskeysierra** [comment]
Just wanted to add to what Thomas wrote. Most use cases that I can observe work in conjunction with Nakadi (our HTTP-based event bus). Typically any background change will be confirmed using an event. So if I place an order (`POST /sales-orders`) I will be able to observe a `sales-order-placed` event that contains the order number from the direct response of the initial request. That makes the HTTP API rather small and simple and also utilizes the event streams that we need to maintain anyway.

**Comment by jmkhael** [comment]
@tfrauenstein thank you for the insights. This was an alternate solution which I am also considering.
Azure's [async operations](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/async-operations) captures it well.

@whiskeysierra that's what I imagined when reading the guidelines, i suspected that you handle this via messaging events. I did not want to force (just yet) our service's clients to also subscribe to an event bus although this aligns better with DDD.
